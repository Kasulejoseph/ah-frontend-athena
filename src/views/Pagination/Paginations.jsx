import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import {
  getArticles,
  getMoreArticles
} from "../../actions/articleActions/ArticleActions";
import Pagination from "../../components/Pagination/Pagination";

export class Paginations extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pageNumber: [],
      currentPage: 1,
      searchParam: ""
    };
  }

  componentWillReceiveProps(nextProps) {
    const {
      moreArticles,
      currentPage,
      fullArticle,
      update,
      searchParam
    } = nextProps;
    this.setState({
      currentPage: currentPage,
      totalPages: fullArticle.total_pages,
      searchParam: searchParam
    });

    if (moreArticles) {
      if (moreArticles.length === 0) {
        update(fullArticle.results);
      } else {
        update(moreArticles);
      }
    }
  }

  handleNext = () => {
    const { currentPage, searchParam } = this.state;
    const { getMoreArticles } = this.props;
    if (searchParam === "") {
      getMoreArticles(currentPage + 1);
    } else {
      getMoreArticles(currentPage + 1, searchParam);
    }
  };

  handlePrevious = () => {
    const { currentPage } = this.state;
    const { getMoreArticles } = this.props;
    getMoreArticles(currentPage - 1);
  };

  render() {
    const { articles, currentPage, totalPages, searchParam } = this.state;
    const { getMoreArticles } = this.props;

    let { pageNumber } = this.state;
    pageNumber = [];
    for (let i = 1; i <= totalPages; i++) {
      pageNumber.push(i);
    }

    return (
      <Pagination
        articles={articles}
        currentPage={currentPage}
        allPages={totalPages}
        nextPage={this.handleNext}
        previousPage={this.handlePrevious}
        pageNumber={pageNumber}
        getMoreArticles={getMoreArticles}
        searchParam={searchParam}
      />
    );
  }
}

Paginations.propTypes = {
  getMoreArticles: PropTypes.func.isRequired,
  moreArticles: PropTypes.shape({}),
  currentPage: PropTypes.shape({}),
  fullArticle: PropTypes.shape({}),
  update: PropTypes.func.isRequired,
  searchParam: PropTypes.shape("")
};

Paginations.defaultProps = {
  moreArticles: {},
  currentPage: {},
  fullArticle: {},
  searchParam: ""
};

export const mapStateToProps = state => {
  return {
    moreArticles: state.articles.articles,
    currentPage: state.articles.currentPage,
    totalPages: state.articles.totalPages,
    searchParam: state.articles.searchParam
  };
};

export default connect(
  mapStateToProps,
  { getMoreArticles, getArticles }
)(Paginations);
