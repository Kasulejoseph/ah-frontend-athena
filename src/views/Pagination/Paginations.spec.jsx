import React from "react";
import { shallow } from "enzyme";
import { Paginations, mapStateToProps } from "./Paginations";

let props = {
  getMoreArticles: jest.fn(),
  pageNumber: [],
  update: jest.fn(),
  fullArticle: {total_pages:1}
};


let wrap;

describe("mocking all functions", () => {
  beforeEach(() => {
    wrap = shallow(<Paginations {...props} />);
  });
  it("should match snapshot", () => {
    expect(wrap).toMatchSnapshot();
  });
  it("should mock handle next", () => {
    wrap.instance().handleNext();
  });
  it("should mock handle previous", () => {
    wrap.instance().handlePrevious();
  });
  it("handleNext with search", () => {
    wrap.setProps({ searchParam: "josephat" });
    wrap.instance().handleNext();
    expect(props.getMoreArticles).toBeCalled();
  });
  it('more articles empty array', () => {
    wrap.setProps({ moreArticles: [] });
    expect(props.update).toBeCalled()
  })
  it('map state to props', () => {
    mapStateToProps({articles:{
      articles: [],
      currentPage: 1,
      totalPages: 3,
      searchParam: ""
    }})
    expect(wrap.instance().props.currentPage).toEqual({})
  })
});
