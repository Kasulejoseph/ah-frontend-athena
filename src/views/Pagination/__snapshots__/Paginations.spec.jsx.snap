// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mocking all functions should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Paginations
    currentPage={Object {}}
    fullArticle={
      Object {
        "total_pages": 1,
      }
    }
    getMoreArticles={[MockFunction]}
    moreArticles={Object {}}
    pageNumber={Array []}
    searchParam=""
    update={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "allPages": undefined,
      "articles": undefined,
      "currentPage": 1,
      "getMoreArticles": [MockFunction],
      "nextPage": [Function],
      "pageNumber": Array [],
      "previousPage": [Function],
      "searchParam": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "allPages": undefined,
        "articles": undefined,
        "currentPage": 1,
        "getMoreArticles": [MockFunction],
        "nextPage": [Function],
        "pageNumber": Array [],
        "previousPage": [Function],
        "searchParam": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
